training/unit-testing/go-lang/README.md

```diff
@@ -0,0 +1,126 @@
+
+### Recommended Go Tutorials
+- [GoConvey Tutorial 1](https://www.smartystreets.com/blog/2015/02/go-testing-part-1-vanillla/)
+- [GoConvey Tutorial 2](https://www.smartystreets.com/blog/2015/02/go-testing-part-2-running-tests/)
+- [GoConvey Tutorial 3](https://www.smartystreets.com/blog/2015/05/go-testing-part-3-convey-behavior/)
+
+### Code Coverage
+When we test certain paths of execution are tested, but often not all.  Knowing which paths are tested is useful, but often the same paths need to be tested multiple time, so 100% coverage doesn't mean 100% tested.
+- [Excellent Golang code coverage walk-through](https://blog.golang.org/cover)
+- [Useful time saving suggestion on stackoverflow](https://stackoverflow.com/a/27284510)
+
+### Assessing How Fast Your Code Runs
+Systems often need to run efficiently, such as a response from a website.  Go incorporates a benchmarking routines, which will time how long the code will take to run.  It clever though as it repeats the tests until it feel it's got an accurate time.  
+- [Downloadable benchmarking example](benchmarking/benchmark_test.go)
+
+###Splitting Up Tests | Setup and Tear Down Tests 
+Its sometimes useful to use a standard way of splitting up tests, or creating a setup and tear down routines for tests.  Golang provides this functionality built in.
+- [Downloadable example without using testing.M](helloworld/split_test.go)
+- [Downloadable example using testing.M](helloworld/split_with_test_main_test.go)
+- [testing.M walk-through](http://cs-guy.com/blog/2015/01/test-main/)
+
+###Running Time Critical Tests
+If your code runs in a higher priority thread at the operating system level, you can also run your tests at this level also: 
+- [Downloadable time critical test example](#time-critical/testing_with_main_OS_thread_test.go)
+
+Table Driven Tests
+------------------
+Go Convey is very useful for creating very readable tests, however occasionally you may need to test lot of different permutations of a similar test, which could result in very long files.  This is where Table Driven Tests are very useful.  Effectively you create a template test, and feed in lots of different test permutations.  Here is a [Table Driven Test example](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)
+
+### Other Interesting Testing Related Information
+- [Dot import for Convey](https://smartystreets.com/blog/2015/05/go-testing-part-3-convey-behavior/)
+
+##Wordier Information
+###Background To Unit Testing
+Unit testing is the process of testing a portion of an application or service to ensure it meets our requirements.  The portion of code is ideally a single function, but is typically less than a few go files.
+
+Ideally we should only test the portion of the application we are focused on, and shouldn't connect with other areas of a system.  This can be difficult, and so sometimes different portions of the system are 'mocked'.
+
+### Is Unit Testing Ever Pointless?
```

Comment:
I wonder if there should be a piece here regarding duplicating tests across functions that wrap one another.

For example a handle method that then calls off to a validate method on an object should only contain 1 or 2 tests to cover for success and failure coming from the validate method even if the validate method contains 8 individual pathways which could lead to failure.

Those 8 individual pathways to failure should be unit tested against the validate method to get close to 100% test coverage.

Side note: if your method or function has so many unit tests against it, there is an argument it should be refactored to reduce complexity, once refactored we can apply the logic in the above paragraph and this should result in clearer unit tests for readability and better coverage.

Reply:
Yip that sounds really good.  Do you happen to have an example I can point to?  Even a before and after?

Reply:
As per direct message the dp-identity-api has been written in a way where unit test coverage is not duplicated across handlers and bespoke methods/functions. Check out the Tokens handler and the ValidateCredentials method with their respective unit tests.